src/libraries/System.Text.Json/docs/ReferenceHandling_spec.md

```diff
@@ -0,0 +1,857 @@
+    "$id": "1",
+    "Manager": {
+        "$ref": "1"
+    }
+}
+```
+
+* `$id` is duplicated (not necessarily nested):
+  * **Newtonsoft.Json**: Throws - Error reading object reference '1'- Inner Exception: ArgumentException: A different value already has the Id '1'.
+  * **S.T.Json**: Throws - Duplicated id found while preserving reference.
+```json
+[
+    {
+        "$id": "1",
+        "Name": "Angela"
+    },
+    {
+        "$id": "1",
+        "Name": "Bob"
+    }
+]
+```
+
+## Preserved arrays
+A regular array is `[ elem1, elem2 ]`.
+A preserved array is written in the next format `{ "$id": "1", "$values": [ elem1, elem2 ] }`
+
+* Preserved array does not contain any metadata:
+  * **Newtonsoft.Json**: Throws - Cannot deserialize the current JSON object into type 'System.Collections.Generic.List`1
+  * **S.T.Json**: Throw - Preserved array $values property was not present or its value is not an array.
+
+  ```json
+  {}
+  ```
+
+* Preserved array only contains $id:
+  * **Newtonsoft.Json**: Throws - Cannot deserialize the current JSON object into type 'System.Collections.Generic.List`1
+  * **S.T.Json**: Throw - Preserved array $values property was not present or its value is not an array.
+
+  ```json
+  {
+      "$id": "1"
+  }
+  ```
+
+* Preserved array only contains `$values`:
+  * **Newtonsoft.Json**: Does not throw and the payload evaluates to the array in the property.
+  * **S.T.Json**: Throw - Preserved arrays cannot lack an identifier.
+
+  ```json
+  {
+      "$values": []
+  }
+  ```
+
+* Preserved array `$values` property is null
+  * **Newtonsoft.Json**: Throw - Unexpected token while deserializing object: EndObject. Path ''.
+  * **S.T.Json**: Throw - Preserved array $values property was not present or its value is not an array.
+
+  ```json
+  {
+      "$id": "1",
+      "$values": null
+  }
+  ```
+
+* Preserved array `$values` property is a primitive value
+  * **Newtonsoft.Json**: Unexpected token while deserializing object: EndObject. Path ''.
+  * **S.T.Json**: Throw - Preserved array $values property was not present or its value is not an array.
+
+  ```json
+  {
+      "$id": "1",
+      "$values": 1
+  }
+  ```
+
+* Preserved array `$values` property contains object
+  * **Newtonsoft.Json**: Unexpected token while deserializing object: EndObject. Path ''.
+  * **S.T.Json**: Throw - Preserved array $values property was not present or its value is not an array.
+
+  ```json
+  {
+      "$id": "1",
+      "$values": {}
+  }
+  ```
+
+## JSON Objects if not Collection (Class | Struct | Dictionary) - On Deserialize (and Serialize?)
+
+* `$ref` **Valid** under conditions:
+  * must be the only property in the object.
+
+* `$id` **Valid** under conditions:
+  * must be the first property in the object
+
+* `$values` **Not Valid**
+
+* `$.*` **Valid**
+
+Note: For Dictionary keys on serialize, should we allow serializing keys `$id`, `$ref` and `$values`? If we allow it, then there is a potential round-tripping issue.
```

Comment:
If the implementation wouldn't otherwise do the string comparisons, I wouldn't introduce this check for essentially every dictionary key.
It seems odd to not serialize given content. Maybe the caller should pay the price of not roundtripping for those cases. Does Newtonsoft.Json allow these keys?

Reply:
As per further discussions, when your enable `Preserve`, properties with leading $ will be forbidden; to workaround that you will need to escape the leading $. 
So we can add an update comment that states that our serializer will never produce a property with leading $ without escaping the $.

Reply:
See resolution a [few lines below](https://github.com/dotnet/runtime/pull/354/files#diff-862e8417d479d30d1ade2dd9acf21333R645) (line 645)